src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,750 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
+using System.Net.Quic;
+using System.IO;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+
+namespace System.Net.Http
+{
+    internal sealed class Http3Connection : HttpConnectionBase, IDisposable
+    {
+        /// <summary>
+        /// If we receive a settings frame larger than this, tear down the connection with an error.
+        /// </summary>
+        private const int MaximumSettingsPayloadLength = 4096;
+
+        /// <summary>
+        /// Unknown frame types with a payload larger than this will result in tearing down the connection with an error.
+        /// Frames smaller than this will be ignored and drained.
+        /// </summary>
+        private const int MaximumUnknownFramePayloadLength = 4096;
+
+        private readonly HttpConnectionPool _pool;
+        private readonly HttpAuthority _origin;
+        private readonly HttpAuthority _authority;
+        private readonly byte[] _altUsedEncodedHeader;
+        private QuicConnection _connection;
+
+        // Keep a collection of requests around so we can process GOAWAY.
+        private readonly Dictionary<long, Http3RequestStream> _activeRequests = new Dictionary<long, Http3RequestStream>();
```

Comment:
For my knowledge, why did you decide not to use a ConcurrentDictionary here?

Reply:
`ConcurrentDictionary` makes more sense if you've got high contention and it's the only change you need to make as one atomic unit. I don't think that was the case here.

That said, I need to revisit locking once I have a baseline. Due to some unknowns at the time of writing, the locks is held uncomfortably long.